### ISSUE TITLE
Compatibility regression with Autodesk Maya in post-rewrite VxKex

### ISSUE THREAD

**Issue by TiberiumFusion** [vxiiduu/VxKex#131]
VxKex 0.0.0.2 works with Maya (2022) / Adsk Licensing v12 without issue. VxKex 1.1.0.1291 unfortunately has problems.

## Old VxKex
With VxKex 0.0.0.2, maya loads just fine, its plugins load just fine, and the right mess of interconnected licensing processes launch all run just fine.

And these are the settings I've been using for the last two years:
![vxkex 0 0 0 2 settings](https://github.com/vxiiduu/VxKex/assets/6332277/14087fc5-4807-4dc0-b845-0578e4709fab)
Wait for child processes is required, because VxKex is needed for maya's renderer plugin, which tries to use NT 6.3 functions. Sometimes Maya runs that plugin as a thread in the maya.exe process, sometimes it runs the plugin in a separate process. In fact, Maya itself doesn't need VxKex at all, only this plugin does.

## New VxKex
With VxKex 1.1.0.1291, maya and its plugins run fine, but all of the autodesk licensing processes now fail to start. As soon as you launch maya, you get a constant spew of 0xc0000142 message boxes from the 3-5 different licensing processes that start and restart multiple times alongside maya.
![spew examples](https://github.com/vxiiduu/VxKex/assets/6332277/50345ab0-f0a7-47a9-8154-47e836afc57e)
Just to be clear, this error does _not_ occur under VxKex 0.0.0.2, and it does _not_ occur when Maya runs without VxKex.

And these are the settings I used:
![vxkex 1 1 0 1291 settings](https://github.com/vxiiduu/VxKex/assets/6332277/2b9f5958-680c-4503-9007-e528b0729556)
- Enabling "Disable all application-specific workarounds" doesn't change anything.
- Enabling "Disable VxKex for child processes" **does technically solve the issue**; the 0xc0000142 errors stop occurring. However, it defeats the point of using VxKex with maya, since I need VxKex to propagate itself to the render plugin processes.

## Notes
I really hope this is not a limitation of the appverifier strategy over the old debugger one.

I mentioned Maya (2022) before in #8, where VxKex 0.0.0.1 was causing "The 32-bit DLL foo was loaded into a 64-bit application" errors from these same autodesk licensing processes. That issue was solved in time for 0.0.0.2 and I never had another any further problems with the licensing processes. Well, until trying the VxKex rewrite. I wonder if this issue is related to that old one.

Also, I must note that there is nothing abnormal about the autodesk licensing executables in this scenario. This is AdskLicensing v12 in particular, and it is not, erm, modified in any way. It runs on my PC no differently from how it runs on the millions of other Autodesk-infected PCs. Oh and just like Maya, the licensing programs don't need VxKex either.

As a workaround to these issues, it would be nice if VxKex had a whitelist and/or blacklist for propagating itself to child processes. With a blacklist, I could add AdskLicensingService.exe, AdskAccessServiceHost.exe, and ADPClientService.exe, but that actually wouldn't fully work, since they spawn various cmd.exe child processes, and blacklisting cmd.exe would be bad. So a whitelist would work better here, where I could whitelist the image name or path of the render plugin child processes.

## Logs
I've attached logs from VxKex's fancy new logging system.
[01 Normal settings launch (with error).zip](https://github.com/vxiiduu/VxKex/files/14629927/01.Normal.settings.launch.with.error.zip)

This first set of logs was generated by launching maya as described in the **New VxKex** section above. I closed 11 of the 0xc0000142 popups during the long maya loading splash before terminating the process. We can see from the logs that all 3 licensing processes appear to fail in the same way.

[02 Disable vxkex for child processes launch (no errors).zip](https://github.com/vxiiduu/VxKex/files/14629929/02.Disable.vxkex.for.child.processes.launch.no.errors.zip)
Log set no.2 was generated by launching maya the same way, except I enabled `Disable VxKex for child processes`, meaning maya launched just fine with no errors from the licensing processes. What is interesting to note is that there is only a single process log file in this set: the one for maya.exe. No log file was created for any of the licensing processes. I am not intimately familiar with exactly what maya is doing when it starts these licensing processes, but each of the licensing processes do typically self-terminate very quickly if a prior process of the same image is already running.

Pass for both zips is "vxkex" (to keep github harvester bots out)

**Comment by vxiiduu** [comment]
This does not seem to be an inherent limitation of the verifier loading method. There seems to be some weirdness in the default loader DLL path of all of the Maya child processes, perhaps Maya is messing with the path environment variable or the environment variables of its child processes. I added some more logging to try and see what is going on, since unfortunately I sort of assumed that that KexpShrinkDllPath function would almost never fail. This build almost certainly will not solve the problem, but try again with it and post the logs.

![image](https://github.com/vxiiduu/VxKex/assets/73044267/755b287f-7ff3-4d33-be13-54d583338ec1)

There should be a new log item like this. I need to see what the original loader path is. VxKex depends on there being duplicate entries in there (for example, C:\Windows and C:\Windows\system32 are both repeated multiple times in the above screenshot) so that it can delete those and add itself in their place.

Anyway, there might be a way for you to fix this problem yourself in the meantime. You mention a whitelist. In fact, this is possible. The way the VxKex options in the properties dialog work is more flexible than you might think. Let's say that you have the following situation:

1. Maya.exe starts child1.exe
2. Maya.exe starts child2.exe
3. Maya.exe starts child3.exe

Imagine you want Maya.exe and child3.exe to run under VxKex, but none of the others. In this situation, you could enable VxKex for Maya with "disable for child processes". This means that Maya will have VxKex enabled, but child1 and child2 will not have VxKex enabled.

Now, how are we going to get VxKex enabled for child3.exe? Just enable VxKex for child3.exe as well in the properties.

Enabling VxKex for a program works no matter how it is started. It can be run by double clicking a shortcut, by typing its name into command prompt, or by being started by another process.

The other settings, like version spoof and disable for child processes, also still apply even if the main "enable vxkex" option is not checked. So if you check the "disable for child processes" option, it will terminate the chain of processes that have VxKex enabled, so to speak. I'm not sure if this explanation is very clear, but I hope you understand nonetheless. After all, there is a reason why not checking "Enable VxKex for this program" doesn't gray out all the other boxes, and that's precisely to allow for flexibility like this.

[Uploading KexSetup_Debug.zip…]()

**Comment by TiberiumFusion** [comment]
The whitelist approach you described makes perfect sense and worked as intended! I was able to work around the problematic licensing programs by only enabling VxKex for the render executables which actually need it. I indeed had been wondering why the advanced options stayed available even when the main VxKex toggle was disabled.

I think VxKex could use a KB somewhere (Wiki tab here on github maybe?) or even just a chm with helpful info and usage instructions for these kind of scenarios.

As for the root problem, I will re-test with your debug build once that's uploaded and get you the logs.

**Comment by vxiiduu** [comment]
Thanks for the reminder, here's that debug build I was going to upload.

[KexSetup_Debug.zip](https://github.com/vxiiduu/VxKex/files/14629113/KexSetup_Debug.zip)

**Comment by TiberiumFusion** [comment]
Logs from repeating the steps in my OP:
[03 Debug build logs (normal settings launch, with error).zip](https://github.com/vxiiduu/VxKex/files/14629541/03.Debug.build.logs.normal.settings.launch.with.error.zip)

**Comment by vxiiduu** [comment]
Yeah, it seems like the duplicate path entries which are usually there do not exist.

In order to fix the problem:

1. In the Control Panel, go to System and Security -> System -> Advanced system settings. This dialog should appear. ![image](https://github.com/vxiiduu/VxKex/assets/73044267/b7cd5105-bd25-4bb5-b9b6-67bb71392376)
2. Click on Environment Variables and scroll down the "System variables" list until you find the one that is titled "Path".
3. Copy and paste "C:\Windows;C:\Windows;C:\Windows;" at the front of the variable value as shown. ![image](https://github.com/vxiiduu/VxKex/assets/73044267/a09ebe18-fa6e-48ad-8518-d2065a81b530)
4. Click OK on all the dialogs and try again.

I don't know why Maya is removing the duplicated entries that are usually there before starting the licensing processes, but that should work around the issue. I'll have to remember to add a better error dialog for this situation since I know now that it's actually possible in the real world.

**Comment by vxiiduu** [comment]
Sorry, ignore my previous post. That wasn't the issue, but now knowing what that DLL path is I can do more testing.

**Comment by TiberiumFusion** [comment]
Ah I was just typing a reply when I saw your edit.

Well, I'll still mention it anyways: Maya isn't removing duplicates from the user or system path. Those two env vars just never had any duplicates in them to begin with. Partly this is because I periodically clean them out on this particular machine, usually after running things that make duplicates or fail to remove old paths, like JDK and powershell installers. And that's because the path on this machine is sometimes too long for some applications.
I'll also note that I have only very rarely seen duplicate entries for windir or system32 on my 7 machines. I've seen a single duplicate of `C:\Windows\system32` before once or twice, but that's it.

Does VxKex really rely on a duplicate entry to work properly? That is interesting. Maybe it'd be better to have the VxKex installer ensure there is a duplicate entry, or otherwise reserve the characters it needs by adding a dummy path like the VxKex install dir.

At any rate, thanks for looking into this issue and I'm happy to gather more test data for you if you need it.

**Comment by vxiiduu** [comment]
The way the loader sets up the path is it adds C:\Windows and C:\Windows\system32 before the Path environment variable. By default when you install Windows, the Path environment variable has C:\Windows and C:\Windows\system32 in it already. That means that by the time VxKex gets to removing the duplicates there should be two of both.

In fact after reading the log a little more closely those duplicates are still there in the Autodesk licensing, but VxKex isn't able to remove them. I think it must be because of a bug in my code rather than anything to do with your system.

**Comment by TiberiumFusion** [comment]
Even more interesting! If you ever write a white paper on the development of VxKex, I'd love to read a section about how you've exploited various strange & ancient decision by the microsoft engineers.

**Comment by vxiiduu** [comment]
Alright, I have fixed the bug.

In case you are curious, the problem happened because you have *forward slashes* in your %PATH% environment variable. I don't know why they are there, but either way they are valid path separators in Windows, so I have added code to VxKex to handle those forward slashes properly. The bug fix will be included in the next release.

**Comment by TiberiumFusion** [comment]
I actually don't have any forward slashes in my system or user path env vars. That is one thing Maya is indeed changing - it is converting all the back slashes into forward slashes. Since maya is also available for unix systems this is probably some common denominator they decided to use, since it also happens to be valid on windows. Other programs might be doing this too, unfortunately.

Thanks for fixing this issue. I'll test it all again in the next release and let you know if something still isn't working right.

**Comment by vxiiduu** [comment]
The recently released 1.1.1.1375 should solve this problem. Feel free to re-open the issue if it still doesn't work.
